### PR TITLE
Added with_temporary_changes to handle updating limit

### PR DESCRIPTION
**Pull request by syncrou**
1. Pass the post url with extra_vars into a block
2. Make a separate patch call to update the limit
3. Yield the original launch restful call
4. Ensure the limit is set back to what it originally was.

### CODE REVIEW

**Comment by syncrou** [comment]
@bdunne, @gmcculloug - Please review - cc @bzwei @mkanoor

**Comment by syncrou** [comment]
@miq-bot add_label enhancement

File: lib/ansible_tower_client/error.rb [hunk 1]
```diff
@@ -32,8 +32,17 @@ def initialize(klass)
       @message = "Error on #{klass.class.name}: #{klass.inspect}"
       super(message)
     end
+
+    def self.parse(text)
+      txt = []
+      JSON.parse(text).each do |name, error|
```
Comment (bzwei) [inline lib/ansible_tower_client/error.rb]:
```
JSON.parse(text).collect do ...
end.join("\n")
```

File: lib/ansible_tower_client/job_template.rb [hunk 2]
```diff
@@ -10,12 +10,19 @@ def initialize(api, raw_body)
...
+      resp = with_temporary_changes(new_limit) do
+        api.post(launch_url, options).body
+      end
+
+      begin
+        job = JSON.parse(resp)
+        api.jobs.find(job['job'])
+      rescue ResourceNotFound
+        Error.parse(resp)
+      end
```
Comment (bzwei) [inline lib/ansible_tower_client/job_template.rb]: What is the return on success and failure? Is it either a `Job` or an error string? An error should be thrown in case of failure.

File: lib/ansible_tower_client/job_template.rb [hunk 3]
```diff
@@ -10,12 +10,19 @@ def initialize(api, raw_body)
       super
     end
 
-    def launch(vars = {})
+    def launch(options = {})
```
Comment (bzwei) [inline lib/ansible_tower_client/job_template.rb]: I would prefer `launch(limit = nil, vars = {})`
If to use options, it should have format `:limit => 'some condition', :extra_vars => {the extra vars}`

File: lib/ansible_tower_client/job_template.rb [hunk 4]
```diff
@@ -27,5 +34,28 @@ def survey_spec
...
+      rescue => e
+        raise InvalidPost.new(e)
+      ensure
+        patch("{ \"limit\": \"#{old_limit}\" }")
+      end
+    end
+
+    def patch(body)
+      api.patch do |req|
+        req.url url
```
Comment (bzwei) [inline lib/ansible_tower_client/job_template.rb]: If we all follow the same style by manageiq, we should write as `req.url(url)`. I am surprised at how to setup the `url` vs `body` as the syntax is different.

File: ansible_tower_client.gemspec [hunk 5]
```diff
@@ -35,4 +35,5 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency "factory_girl"
   spec.add_development_dependency "rake", "~> 10.0"
   spec.add_development_dependency "rspec"
+  spec.add_development_dependency "byebug"
```
Comment (bdunne) [inline ansible_tower_client.gemspec]: Please do this in a separate PR.

File: lib/ansible_tower_client/job_template.rb (hunk 3 above)
Comment (bdunne) [inline lib/ansible_tower_client/job_template.rb]: @bzwei I suggested `options` due to the number of optional args that could be passed in here.  In the future we'll also want to pass in survey data.

File: lib/ansible_tower_client/job_template.rb [hunk 6]
```diff
@@ -10,12 +10,19 @@ def initialize(api, raw_body)
...
     end
 
-    def launch(vars = {})
+    def launch(options = {})
       launch_url = "#{url}launch/"
-      extra = JSONValues.new(vars).extra_vars
-      resp = api.post(launch_url, extra).body
-      job = JSON.parse(resp)
-      api.jobs.find(job['job'])
+      new_limit = options.delete('limit')
```
Comment (bdunne) [inline lib/ansible_tower_client/job_template.rb]: @syncrou we may want to `options.dup` to prevent modifying the real options

File: lib/ansible_tower_client/job_template.rb [hunk 7]
```diff
@@ -10,12 +10,19 @@ def initialize(api, raw_body)
...
+      resp = with_temporary_changes(new_limit) do
```
Comment (bdunne) [inline lib/ansible_tower_client/job_template.rb]: No need to shorten `response`

File: lib/ansible_tower_client/job_template.rb [hunk 8]
```diff
@@ -10,12 +10,19 @@ def initialize(api, raw_body)
...
+        api.post(launch_url, options).body
+      end
+
+      begin
+        job = JSON.parse(resp)
+        api.jobs.find(job['job'])
+      rescue ResourceNotFound
```
Comment (bdunne) [inline lib/ansible_tower_client/job_template.rb]: Did we not hit this error in the past?

Comment (bdunne) [inline lib/ansible_tower_client/job_template.rb]: This type of exception handling should probably be on the collection, not here.

File: lib/ansible_tower_client/job_template.rb (hunk 2 above)
Comment (bdunne) [inline lib/ansible_tower_client/job_template.rb]: Agreed, object or raise

File: lib/ansible_tower_client/job_template.rb [hunk 9]
```diff
@@ -27,5 +34,28 @@ def survey_spec
...
+
+    private
+
+    def with_temporary_changes(limit)
+      old_limit = @raw_body['limit']
+      new_limit = limit
+      patch("{ \"limit\": \"#{new_limit}\" }")
+      begin
+        yield
+      rescue => e
```
Comment (bdunne) [inline lib/ansible_tower_client/job_template.rb]: Don't rescue all exceptions.  I expect that any errors would be from the `api.post` [here](https://github.com/ManageIQ/ansible_tower_client/pull/22/files#diff-c799bcb40bc98d6ec3bcd2e4b41361e6R17) and only those should be rescued.

File: lib/ansible_tower_client/job_template.rb (hunk 8 above)
Comment (syncrou) [inline lib/ansible_tower_client/job_template.rb]: @bdunne - We did hit this in the past - I was trying to handle it better with the `Error.parse` method. In the past Faraday would raise ResourceNotFound - but the actual text of the that error - e.g. the API response was lost.  If I'm reading what you and @bzwei want - we still need to raise an error - but pass the API response as text shown in the error class message?

File: lib/ansible_tower_client/job_template.rb (hunk 9 above)
Comment (syncrou) [inline lib/ansible_tower_client/job_template.rb]: I was going to ask about that - I didn't feel great about leaving `rescue` open ended either but wasn't sure on the standard operating procedure.

**Comment by syncrou** [comment]
@bdunne - @bzwei - I applied the requested changes - Please review.

File: lib/ansible_tower_client/error.rb [hunk 10]
```diff
@@ -28,12 +28,20 @@ def general_msg
 
   class Error < StandardError
     attr_reader :message
-    def initialize(klass)
-      @message = "Error on #{klass.class.name}: #{klass.inspect}"
-      super(message)
+    def initialize(klass, message = nil)
+      @message = message.nil? ? "Error on #{klass.class.name}: #{klass.inspect}" : message
```
Comment (bdunne) [inline lib/ansible_tower_client/error.rb]: Maybe `@message = message || "Error on #{klass.class.name}: #{klass.inspect}"`

File: lib/ansible_tower_client/job_template.rb [hunk 11]
```diff
@@ -10,11 +10,15 @@ def initialize(api, raw_body)
       super
     end
 
-    def launch(vars = {})
+    def launch(options = {})
       launch_url = "#{url}launch/"
-      extra = JSONValues.new(vars).extra_vars
-      resp = api.post(launch_url, extra).body
-      job = JSON.parse(resp)
+      new_limit = options.dup.delete('limit')
```
Comment (bdunne) [inline lib/ansible_tower_client/job_template.rb]: I think this needs to be:
```
options = options.dup
new_limit = options.delete('limit')
```

File: lib/ansible_tower_client/job_template.rb [hunk 12]
```diff
@@ -10,11 +10,15 @@ def initialize(api, raw_body)
...
+      response = with_temporary_changes(new_limit) do
+        api.post(launch_url, options).body
+      end
+
+      job = JSON.parse(response)
+      raise InvalidOptions.new(self, Error.parse(response)) if job['job'].nil?
```
Comment (bdunne) [inline lib/ansible_tower_client/job_template.rb]: I still don't know why we need this since we are handling this error [here](https://github.com/ManageIQ/ansible_tower_client/blob/master/lib/ansible_tower_client/collection.rb#L27)

Comment (syncrou) [inline lib/ansible_tower_client/job_template.rb]: The thought was so we could see errors like this `AnsibleTowerClient::InvalidOptions: Error on variables_needed_to_start: ["'instance_ids' value missing", "'crap_id' value missing"]`

File: lib/ansible_tower_client/job_template.rb (hunk 11 above)
Comment (bdunne) [inline lib/ansible_tower_client/job_template.rb]: In this implementation, we still pass limit in the post options on line 17.  That was supposed to be removed from the options since the block already takes care of setting the limit for us.

```ruby
irb(main):001:0> raw_options = {:limit => "abc", :extra_vars => "xyz"}
=> {:limit=>"abc", :extra_vars=>"xyz"}
irb(main):002:0> options = raw_options
=> {:limit=>"abc", :extra_vars=>"xyz"}
irb(main):003:0> raw_options.object_id
=> 47375013423080
irb(main):004:0> options.object_id
=> 47375013423080
irb(main):005:0> limit = options.dup.delete(:limit)
=> "abc"
irb(main):006:0> options
=> {:limit=>"abc", :extra_vars=>"xyz"}
```

The `dup` needs to be on its own line because:
```ruby
irb(main):001:0> raw_options = {:limit => "abc", :extra_vars => "xyz"}
=> {:limit=>"abc", :extra_vars=>"xyz"}
irb(main):002:0> options = raw_options.dup
=> {:limit=>"abc", :extra_vars=>"xyz"}
irb(main):003:0> raw_options.object_id
=> 47002009217800
irb(main):004:0> options.object_id
=> 47002008338660
irb(main):005:0> limit = options.delete(:limit)
=> "abc"
irb(main):006:0> limit
=> "abc"
irb(main):007:0> options
=> {:extra_vars=>"xyz"}
irb(main):008:0> raw_options
=> {:limit=>"abc", :extra_vars=>"xyz"}
```

File: lib/ansible_tower_client/job_template.rb (hunk 12 above)
Comment (bdunne) [inline lib/ansible_tower_client/job_template.rb]: Thanks, the discussion cleared up a lot.  You're correct, this is a different problem from the resource not being found.
I was expecting Faraday to raise an error when it receives a HTTP 400 response from the `api.post` above.  I think we should probably handle this error generically in the Api class.  Let's do that in a followup PR.

**Comment by syncrou** [comment]
@bdunne - Pushed those changes

**Comment by bdunne** [comment]
:+1:

**Comment by syncrou** [comment]
@bdunne - rebased
